pkg/urbit/vere/io/ames.c

```diff
@@ -324,144 +325,143 @@ _ames_ship_of_chubs(c3_d sip_d[2], c3_y len_y, c3_y* buf_y)
   memcpy(buf_y, sip_y, c3_min(16, len_y));
 }
 
-/* _ames_sift_prelude(): parse prelude, returning length
+/* _ames_sift_head(): parse packet header.
 */
-static c3_y
-_ames_sift_prelude(u3_head* hed_u,
-                   u3_prel* pre_u,
-                   c3_w     len_w,
-                   c3_y*    pre_y)
+static void
+_ames_sift_head(u3_head* hed_u, c3_y buf_y[4])
 {
-  c3_y rog_y, sen_y, rec_y, len_y;
+  c3_w hed_w = _ames_sift_word(buf_y);
 
-  rog_y = ( c3y == hed_u->rel_o )? 6 : 0;
-  sen_y = 2 << hed_u->sac_y;
-  rec_y = 2 << hed_u->rac_y;
-  len_y = rog_y + sen_y + rec_y + 1;
+  //  first three bits are reserved
+  //
+  hed_u->req_o = (hed_w >>  2) & 0x1;
+  hed_u->sim_o = (hed_w >>  3) & 0x1;
+  hed_u->ver_y = (hed_w >>  4) & 0x7;
+  hed_u->sac_y = (hed_w >>  7) & 0x3;
+  hed_u->rac_y = (hed_w >>  9) & 0x3;
+  hed_u->mug_l = (hed_w >> 11) & 0xfffff; // 20 bits
+  hed_u->rel_o = (hed_w >> 31) & 0x1;
+}
 
-  pre_u->sic_y = pre_y[0] & 0xf;
-  pre_u->ric_y = pre_y[0] & 0xf0;
+/* _ames_sift_prel(): parse prelude,
+*/
+static void
+_ames_sift_prel(u3_head* hed_u,
+                u3_prel* pre_u,
+                c3_y*    buf_y)
+{
+  c3_y sen_y, rec_y;
+  c3_w cur_w = 0;
+
+  pre_u->sic_y = buf_y[0] & 0xf;
+  pre_u->ric_y = buf_y[0] & 0xf0;
+  cur_w++;
 
-  _ames_ship_to_chubs(pre_u->sen_d, sen_y, pre_y + 1);
-  _ames_ship_to_chubs(pre_u->rec_d, rec_y, pre_y + 1 + sen_y);
+  sen_y = 2 << hed_u->sac_y;
+  _ames_ship_to_chubs(pre_u->sen_d, sen_y, buf_y + cur_w);
+  cur_w += sen_y;
 
-  if (rog_y) {
+  rec_y = 2 << hed_u->rac_y;
+  _ames_ship_to_chubs(pre_u->rec_d, rec_y, buf_y + cur_w);
+  cur_w += rec_y;
+
+  if ( c3y == hed_u->rel_o ) {
     c3_y rag_y[8] = {0};
-    memcpy(rag_y, pre_y + 1 + sen_y + rec_y, rog_y);
+    memcpy(rag_y, buf_y + cur_w, 6);
```

Comment:
Why 6? If it's the size of a particular type, we should use `sizeof(<type>)`. If it's not, we should create a named constant.